sdk-exporter/prometheus/src/main/scala/org/typelevel/otel4s/sdk/exporter/prometheus/PrometheusWriter.scala

```diff
@@ -0,0 +1,437 @@
+import org.typelevel.otel4s.sdk.exporter.prometheus.PrometheusMetricExporter.Defaults
+import org.typelevel.otel4s.sdk.exporter.prometheus.PrometheusTextRecord.PrometheusTextPoint
+import org.typelevel.otel4s.sdk.metrics.data.AggregationTemporality
+import org.typelevel.otel4s.sdk.metrics.data.MetricData
+import org.typelevel.otel4s.sdk.metrics.data.MetricPoints
+import org.typelevel.otel4s.sdk.metrics.data.PointData
+import org.typelevel.otel4s.sdk.metrics.data.PointData.DoubleNumber
+import org.typelevel.otel4s.sdk.metrics.data.PointData.LongNumber
+
+import scala.collection.immutable.ListMap
+
+trait PrometheusWriter[F[_]] {
+  def contentType: String
+  def write[G[_]: Foldable](metrics: G[MetricData]): Stream[F, Byte]
+}
+
+object PrometheusWriter {
+
+  private val TargetInfoName = "target_info"
+  private val TargetInfoDescription = "Target metadata"
+
+  private val ScopeInfoMetricName = "otel_scope_info"
+  private val ScopeInfoDescription = "Instrumentation Scope metadata"
+
+  private val ScopeInfoNameLabel = "otel_scope_name"
+  private val ScopeInfoVersionLabel = "otel_scope_version"
+
+  private val Counter = "counter"
+  private val Gauge = "gauge"
+  private val Histogram = "histogram"
+
+  private val PosInf = "+Inf"
+  private val NegInf = "-Inf"
+
+  private final case class MetricGroup(
+      prometheusName: String,
+      prometheusType: String,
+      description: Option[String],
+      metrics: NonEmptyVector[MetricData]
+  ) {
+    def helpLine(metricName: String): Option[String] = description.map { h =>
+      s"$metricName ${escapeString(h)}"
+    }
+  }
+
+  private final case class MetricAggregate(
+      groups: Map[String, MetricGroup] = Map.empty,
+      scopes: Map[String, InstrumentationScope] = Map.empty,
+      resource: Option[TelemetryResource] = None
+  ) {
+    def tryAddScope(scope: InstrumentationScope, disableScopeInfo: Boolean): MetricAggregate = {
+      if (scope.attributes.nonEmpty && !disableScopeInfo) {
+        this.copy(scopes = this.scopes + (scope.name -> scope))
+      } else {
+        this
+      }
+    }
+
+    def tryUpdateResource(resource: TelemetryResource, disableTargetInfo: Boolean): MetricAggregate = {
+      this.resource match {
+        case Some(_)                   => this
+        case None if disableTargetInfo => this
+        case _                         => this.copy(resource = resource.some)
+      }
+    }
+
+    def addOrUpdateGroup(prometheusName: String, group: MetricGroup): MetricAggregate = {
+      this.copy(groups = this.groups + (prometheusName -> group))
+    }
+  }
+
+  def text[F[_]: MonadThrow: Console](
+      withoutUnits: Boolean = Defaults.WithoutUnits,
+      withoutTypeSuffixes: Boolean = Defaults.WithoutTypeSuffixes,
+      disableScopeInfo: Boolean = Defaults.DisableScopeInfo,
+      disableTargetInfo: Boolean = Defaults.DisableTargetInfo
+  ): PrometheusWriter[F] =
```

Comment:
Default values may cause binary compatibility issues in the future if we decide to extend the method. 

- https://docs.scala-lang.org/overviews/core/binary-compatibility-for-library-authors.html#evolving-code-without-breaking-binary-compatibility
- https://github.com/jatcwang/binary-compatibility-guide?tab=readme-ov-file#dont-adding-parameters-with-default-values-to-methods


What if we replace multiple parameters with the config? 
```scala
object PrometheusWriter {

  sealed trait Config {
    def withUnits: Config
    def withoutUnits: Config
    ... 
  }

  object Config {
    def default: Config = ConfigImpl(...)
  
    private final case class ConfigImpl(...) extends Config
  }

  def text[F[_]: MonadThrow: Console](config: Config): PrometheusWriter[F] = ???
}

```

We can also reference this config in different places (e.g. `PrometheusHttpRoutes`):
```scala
object PrometheusHttpRoutes {
  def routes[F[_]: MonadThrow: Console](exporter: MetricExporter.Pull[F]): HttpRoutes[F] = 
    routes(exporter, PrometheusWriter.Config.default)

  def routes[F[_]: MonadThrow: Console](
      exporter: MetricExporter.Pull[F],
      writerConfig: PrometheusWriter.Config
  ): HttpRoutes[F] = ???
}
```

Reply:
I have mixed feelings about `withWithoutUnits`, `withWithoutTypeSuffixes`, etc. 
While it matches OtelJava API, we could offer clearer encoding. 

### 1) Hide arguments entirely:
```scala
def withUnits: HttpServerBuilder[F] = ...
def withoutUnits: HttpServerBuilder[F] = ...
def disableScopeInfo: HttpServerBuilder[F] = ...
def enableScopeInfo: HttpServerBuilder[F] = ...
```

### 2) Use ADT instead (somewhat verbose, but easier to extend):
```scala
sealed trait UnitSuffix
object UnitSuffix {
  def include: UnitSuffix = Include
  def exclude: UnitSuffix = Exclude
  private[prometheus] case object Include extends UnitSuffix
  private[prometheus] case object Exclude extends UnitSuffix
}

sealed trait TypeSuffix
object TypeSuffix {
  def include: TypeSuffix = Include
  def exclude: TypeSuffix = Exclude
  ...
}

sealed trait ScopeInfo 
object ScopeInfo {
  def enable: ScopeInfo = ...
  def disable: ScopeInfo = ...
}

sealed trait Config {
  def withUnitSuffix(suffix: UnitSuffix): Config = ...
  def withTypeSuffix(suffix: TypeSuffix): Config = ...
  def withScopeInfo(info: ScopeInfo): Config = ...
}
```

WDYT?

Reply:
Thanks for the interesting information about backward compatibility. Yes, let's introduce `PrometheusWriter.Config` instead of default parameters.
What about approach with ADT for builder, I'm not sure it will pay off because it doesn't look like we'll add and remove new config options very often.
I think your proposed approach with `withWriterConfig` is the most concise and simplest. That's what I have chosen to use.
Feel free to reopen this thread or create a new one, if there are any other comments.